### PR TITLE
Parameterized retry

### PR DESCRIPTION
**Pull request by lvangool**
To address: https://github.com/fog/fog-aws/issues/448
#1 `retry_if_over_limit` to disable retries fully
#2 `retry_jitter_magnitude` to provide a magnitude to jitter (default same)
#3 replace `splin lock` with `sleep`

### CODE REVIEW

File: lib/fog/aws/compute.rb [hunk 1]
```diff
@@ -551,6 +551,8 @@ def initialize(options={})
           @instrumentor           = options[:instrumentor]
           @instrumentor_name      = options[:instrumentor_name] || 'fog.aws.compute'
           @version                = options[:version]     ||  '2016-11-15'
+          @retry_if_over_limit    = options.has_key?(:retry_if_over_limit) ? options[:retry_if_over_limit] : true
```
Comment (geemus) [inline lib/fog/aws/compute.rb]: I wonder if a clearer name might be `retry_request_limit_exceeded`? What do you think?

File: lib/fog/aws/compute.rb [hunk 2]
```diff
@@ -633,18 +635,12 @@ def _request(body, headers, idempotent, parser, retries = 0)
...
                   Fog::Compute::AWS::NotFound.slurp(error, match[:message])
-                when 'RequestLimitExceeded'
-                  if retries < max_retries
-                    jitter = rand(100)
-                    waiting = true
-                    start_time = Time.now
-                    wait_time = ((2.0 ** (1.0 + retries) * 100) + jitter) / 1000.0
+                when 'RequestLimitExceeded'                  
+                  if @retry_if_over_limit && retries < max_retries
+                    jitter = rand * 10 * @retry_jitter_magnitude
```
Comment (geemus) [inline lib/fog/aws/compute.rb]: The reworking of this math seems to not want to work out in my brain, so I just want to double check I'm interpreting correctly.

I believe in both cases, it should end up that jitter adds some amount between 0 and 1 second to the resulting value (with the default magnitude). The before/after just change where the multiplication/divisions happen to reach this result. Does that match with your understanding/intention?

File: lib/fog/aws/compute.rb (hunk 1 above)
Comment (lvangool) [reply lib/fog/aws/compute.rb]: Yup, happy with `retry_request_limit_exceeded` - got hit by: https://martinfowler.com/bliki/TwoHardThings.html

File: lib/fog/aws/compute.rb (hunk 2 above)
Comment (lvangool) [reply lib/fog/aws/compute.rb]: You're right - 
Original code - the jitter diff 0.01 - 0.1 secs (midpoint 0.05 secs)
An new code - the jitter diff 0.1 - 1 secs (midpoint 0.5 secs) (default)

This makes a bit more sense IMHO given AWS throttling happens at second granularity. 
But happy to change the default to maintain consistency if you prefer?

Ps. I tried to lay it out in such a way that the magnitude makes sense. ie. 
magnitute 0.01 --> jitter midpoint 0.05 secs
magnitute 0.1 (default) --> jitter midpoint 0.5 secs
magnitute 1 --> jitter midpoint 5 secs
magnitute 10 (default) --> jitter midpoint 50 secs
etc.

Comment (geemus) [reply lib/fog/aws/compute.rb]: Thanks for laying that out, I thought there might be a difference in values, but my brain was really not cooperating in the specifics. That seems fine, thanks for the detailed explanation.

**Comment by lvangool** [comment]
Thanks :)

**Comment by lvangool** [comment]
@geemus any plans to cut a new gem version with this? :)

File: lib/fog/aws/compute.rb [hunk 3]
```diff
@@ -546,11 +546,13 @@ class Real
...
-          @region                 = options[:region] ||= 'us-east-1'
-          @instrumentor           = options[:instrumentor]
-          @instrumentor_name      = options[:instrumentor_name] || 'fog.aws.compute'
-          @version                = options[:version]     ||  '2016-11-15'
+          @connection_options           = options[:connection_options] || {}
+          @region                       = options[:region] ||= 'us-east-1'
+          @instrumentor                 = options[:instrumentor]
+          @instrumentor_name            = options[:instrumentor_name] || 'fog.aws.compute'
+          @version                      = options[:version]     ||  '2016-11-15'
+          @retry_request_limit_exceeded = options.has_key?(:retry_request_limit_exceeded) ? options[:retry_request_limit_exceeded] : true
```
[the remark this answers is not in the document]
Comment (lvangool) [reply lib/fog/aws/compute.rb]: 👍

[the remark this answers is not in the document]
Comment (lvangool) [reply lib/fog/aws/compute.rb]: PR for that here: https://github.com/fog/fog-aws/pull/457

**Comment by geemus** [comment]
@lvangool releases are, more often than not, just when somebody asks nicely. So we can certainly try to get that done soon.

@lanej would you be up for cutting the release? My only concern would be that we had some help on improving the changelog, but I think the release process might have scripting in place that would clobber those manual changes? Anyway, would appreciate whatever insights you might have about that, even if you need me to cut the actual release.

**Comment by lvangool** [comment]
Thanks @geemus much appreciated.

**Pull request merged**